core/src/main/scala/org/apache/spark/rpc/netty/Dispatcher.scala

```diff
@@ -68,11 +66,27 @@ private[netty] class Dispatcher(nettyEnv: NettyRpcEnv, numUsableCores: Int) exte
       if (stopped) {
         throw new IllegalStateException("RpcEnv has been stopped")
       }
-      if (endpoints.putIfAbsent(name, getMessageLoop(name, endpoint)) != null) {
+      if (endpoints.containsKey(name)) {
         throw new IllegalArgumentException(s"There is already an RpcEndpoint called $name")
       }
+      val msgLoop = findMessageLoop(name, endpoint)
+      endpoints.put(name, msgLoop)
+      try {
+        endpointRefs.put(endpoint, endpointRef)
```

Comment:
Hmm. This is all correct but feels a bit overkill. Seems like a simpler version would be:

```
endpointRefs.put(endpoint, endpointRef)
try {
  endpoints.put(name, getMessageLoop(name, endpoint))
} catch {
  // cleanup endpointRefs
}
```

Yes, that uses the old `getMessageLoop()` (which could be inlined here for clarity), but that's ok as long as it's done after the `containsKey` check. Then you don't even need the changes to the other file.

Reply:
Just to confirm, that's pretty much close with the patch before ac10f87 (that commit was to reflect review comment), with the additional changes; to use containsKey/put instead of putIfAbsent, and inline getMessageLoop. Could you confirm?

Reply:
Hmm, I can't find a wait to see the complete patch at a specific commit in the UI, so I'll say "maybe".

The goal is:
- not modify "endpoints" when checking if the endpoint exists
- update "endpointRefs" before registering the endpoint's message loop (calling `register` in the case of the shared loop, or creating the dedicated message loop)

Reply:
Thanks for confirming! Will make a change.

Reply:
Addressed.